[Wei Qing] iP

weiquu opened the pull request:
<h1> DukePro </h1>

> "Your mind is for having ideas, not holding them." - David Allen ([source](https://dansilvestre.com/productivity-quotes))

DukePro frees your mind of having to remember things you need to do. It's,
* text-based
* easy to learn
* ~~FAST~~ _SUPER_ FAST to use

All you need to do is,
1. download it from [here](https://nus-cs2103-ay2122s1.github.io/website/schedule/week4/project.html).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you :wink:

And it is __FREE__!

Features:
- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

---

If you're a Java programmer, you can use it to practice Java too. Here's the `main` method:

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

weiquu committed:
Completed Level-1: Greet, Echo, Exit

weiquu committed:
Completed Level-2: Add, List

weiquu committed:
Completed Level-3: Mark as Done

weiquu committed:
Completed Level-4: ToDos, Events, Deadlines

weiquu committed:
Completed A-TextUiTesting: Automated Text UI Testing

weiquu committed:
Completed Level-5: Handle Errors

weiquu committed:
Completed Level-6: Delete

weiquu committed:
A-Enums

weiquu committed:
Refactoring

weiquu committed:
Completed Level-7: Save

weiquu committed:
Completed Level-8: Dates and Times

weiquu committed:
Merge branch 'branch-Level-8'

weiquu committed:
Completed Level-8: Dates and Times

weiquu committed:
Merge branch 'branch-Level-7'

weiquu committed:
Merge branch 'branch-Level-8'

weiquu committed:
Refactor code to be more OOP

Added a Ui class that deals with interactions with the user.
Added a Storage class that deals with loading tasks from the file and
saving tasks in the file.
Added a Parser class that deals with making sense of the user command.
Added a TaskList class that contains the task list.

weiquu committed:
Organise classes into packages

Put all classes into one package, duke, and have another package within
that called task for all classes related to tasks.

weiquu committed:
Add JUnit tests

JUnit tests added for all public methods of all tasks (i.e. ToDo, Event
, and Deadline). Testing of invalid date and/or time for Deadline.

weiquu committed:
Add JavaDoc comments to all classes.
JavaDoc comments added to all methods and members of the classes.

weiquu committed:
Tweak code to comply with coding standard.

Changed small bits of code to comply with the CS2103 coding standard.

weiquu committed:
Add find functionality.

Gives users a way to find a task by searching for a keyword.

weiquu committed:
Merge branch 'branch-A-JavaDoc'

weiquu committed:
Merge branch 'branch-A-CodingStandard'

weiquu committed:
Merge branch 'branch-Level-9'

benluiwj commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,177 @@
+     * Constructs a new instance of the TaskList with no tasks inside.
+     */
+    public TaskList() {
+        this.userTasks = new ArrayList<>();
```
Perhaps including the type of array list you're initialising here. `new ArrayList<Task>()` instead of `new ArrayList<>()`

benluiwj commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,111 @@
+        }
+
+        // Handle representation of information based on whether there is a recognisable date and time
+        int startIndex = hasTime ? 2 : hasDate ? 1 : 0;
```
Try to avoid using magic numbers here :)

benluiwj reviewed (commented):
Good job 👍 
Some minor improvements (that is good to include) requested

ChuangZheQuan commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,87 @@
+import java.util.List;
+
+public class Parser {
+    /** Regex for a Done command. */
```
I don't think you need to put the ```.``` for class members
Taken from the notes:
```/** Number of connections to this database */
private int connectionCount;

ChuangZheQuan commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,111 @@
+    }
+
+    /**
+     * {@inheritDoc}
```
Is ```@inheritDoc``` is part of the coding standard?

weiquu replied:
Yup wasn't really sure about this but my TA mentioned that there was a way to inherit the JavaDoc comments from superclass methods, so I googled and found this!

ChuangZheQuan reviewed (commented):
Very well done :) Had a hard time nitpicking, let alone spotting actual mistakes!

weiquu committed:
Merge commit 'd8398594b7bc43da5eb865321c5a50cec4b3923d' into branch-A-Gradle

weiquu committed:
Add Gradle support to project

Added Gradle support to the project, set up Gradle to run unit tests.

weiquu committed:
Add CheckStyle to detect coding style violations

Set up CheckStyle file (and suppressions file), ran checks on each file
and corrected and coding style violations.

weiquu committed:
Add basic GUI to Duke

Based the GUI off the tutorial.

weiquu committed:
Update GUI for Duke

Components are heavily dependent on each other, and the process of
visually enhancing the GUI is long and painful. In addition, the code is
untidy and long.

Let's refactor the code to use FXML to define our user interface, so
that it is easier to rebuild and control the UI elements in Duke.

weiquu committed:
Add a GUI to Duke

Duke relies on a CLI which is not as graphically appealing as a GUI.
Let's use JavaFX to create a better GUI that retains the same
functionality as Duke while enhancing the user experience.

Merge branch 'branch-Level-10'

weiquu committed:
Link Duke's GUI to main program logic

Program logic is not linked to the input entered by the user in Duke's
GUI. The program also does not exit when the user enters 'bye'. This
needs to be updated in order for Duke to becoming fully functional and
have the GUI act as a good representation of the previous CLI.

Let's link the user input in the GUI to Duke's main logic. A new
Enumeration class, DukeStatus, is used to encapsulate the current status
of Duke (e.g. whether there is an error or should exit) as well as the
response that the user should see in the GUI.

Using enumerations is preferable here in order to ensure that only
certain values are allowed, and to encapsulate both the current state of
Duke as well as it's response in a single class.

weiquu committed:
Add assertions to document assumptions

No assertions are used to ensure that assumptions should hold at various
points in the code. This should change assertions have low impact on
code yet provide a lot of safety.

Let's add assertions to points in the code where certain assumptions
should hold, like to ensure that the tasks have been correctly read from
the data stored in the user's computer.

weiquu committed:
Improve code quality

Code has some quality issues like magic numbers, weak SLAP, long
methods, deep nesting, etc. These affect the readability of the code and
make it harder for someone else to understand.

Let's refactor the code to improve code quality where necessary. In
particular, the code avoid magic numbers, long methods, and deep nesting
while ensuring that it SLAPs hard.

Note that the Parser class will be refactored in a future commit.

weiquu committed:
Use Streams

Double check to ensure that the conversion of data from the user's local
storage to Duke and back makes use of streams correctly.

weiquu committed:
Merge pull request #1 from weiquu/branch-A-Assertions

Add assertions to document assumptions

weiquu committed:
Merge branch 'master' into branch-A-CodeQuality

weiquu committed:
Merge branch 'master' into branch-A-Streams

weiquu committed:
Merge pull request #2 from weiquu/branch-A-CodeQuality

Improve code quality

weiquu committed:
Merge branch 'master' into branch-A-Streams

weiquu committed:
Merge pull request #3 from weiquu/branch-A-Streams

Use Streams

weiquu committed:
Set up CI

Add .yml file so that GitHub Actions can go ahead with Continuous
Integration.

weiquu committed:
Minor refactoring of UI class

weiquu committed:
Add Help function

Duke has no help function to provide in-App guidance to users should
they be unfamiliar with the application. This is an issue as there will
be higher barriers to entry for new users.

Let's add a help function that lets users see all the commands they can
use within the application. This will significantly reduce the barriers
of entry for new users while simulataneously giving more experienced
users the opportunity to remind themselves of the commands and format of
the commands that they can use.

weiquu committed:
Merge pull request #4 from weiquu/branch-BCD-Extension

Add Help function

weiquu committed:
Minor change to GUI

Attempt to change background color of Duke to something more vibrant.

weiquu committed:
Add screenshot and brief User Guide

Duke has no representative screenshot and User Guide for the user to
view.

Let's add these so that the user can go through them and gain a better
understanding of how to use Duke.

weiquu committed:
Tidy up implementation and work on GUI

GUI is very basic and not customised to the implementation of Duke
envisioned. Some methods are also too long and should be separated out
into other smaller, more self-contained methods.

Let's put in new background and user and duke images into the GUI, as
well as set backgrounds for the dialog boxes for the text to contrast
more against the overall background.

Let's also split the method in the Parser class into smaller methods so
that it complies with the coding conventions and is easier for readers
to understand.

weiquu committed:
Insert JavaDoc comments and other GUI changes

JavaDoc comments are out of date and some methods do not have JavaDoc
comments. In addition, the text for the GUI is a bit plain.

Let's add JavaDoc comments to the methods that don't have them and
update the other JavaDoc comments so that the code is understandable to
readers.

Let's also update the text and font of the GUI to make it a bit more
customised to the Duke program.

weiquu committed:
Add a 2 second delay after the bye command

Duke does not have a delay after the bye command is executed, which
means that the program closes immediately after the user sends the
command. It would feel smoother if there was a slight delay.

Let's add a 2 second delay after the bye command is sent to achieve the
purpose stated above.

weiquu committed:
Update README and Ui image

README and Ui image are outdated.

Let's re-write and re-take the image such that they are updated and
reflect the current status of the Duke program.
